[4.0][webservice][com_installer] "list of installed extensions" webservice

stefanoel opened the pull request:
Webservice that lists installed extensions, based on com_installer Manage model

Pull Request for Issue # .

### Summary of Changes
Added a new plugin in /plugins/webservice/installer
Added language files for the plugin

Fixed /administrator/components/com_installer/Model/ManageModel.php that was checking for empty values using !="" instead of empty(), causing an internal error.

Added Controller and View in /api/components/com_installer/


### Testing Instructions
GET http://<yoursite>/api/index.php/v1/installer/manage

### Expected result
A json response with a list of installed extensions like this:
 "data": [
    {
      "type": "manage",
      "id": "160",
      "attributes": {
        "name": "Action Log - Joomla",
        "type": "plugin",
        "folder": "actionlog",
        "client_id": 0,
        "status": 1,
        "version": "3.9.0",
        "id": 160
      }
    },
    {
      "type": "manage",
      "id": "82",
      "attributes": {
        "name": "Action Logs - Latest",
        "type": "module",
        "folder": "",
        "client_id": 1,
        "status": 1,
        "version": "3.9.0",
        "id": 82
      }
    },
...


### Actual result
New feature


### Documentation Changes Required
yes

stefanoel committed:
[4.0] Adding "list of extensions" webservice (com_installer)

Webservice that lists installed extensions based on com_installer Manage model

SharkyKZ commented on administrator/components/com_installer/Model/ManageModel.php:
```diff
@@ -324,7 +324,7 @@ protected function getListQuery()
 		$core     = $this->getState('filter.core');
 
-		if ($status !== '')
+		if (!empty($status))
```
Please revert changes in this file.

SharkyKZ replied:
>Fixed /administrator/components/com_installer/Model/ManageModel.php that was checking for empty values using !="" instead of empty(), causing an internal error.

Hm, can you elaborate on this? The changes introduce bugs because some fields have `0` value option. So `empty()` is not suitable.

stefanoel replied:
ok, I hope the code is now fixed... sorry, this is my first experience with Github, now I'm going to fix some code styling error
thx!

SharkyKZ commented:
The XML is shown as binary. You may need to configure your editor to store files as ASCII.

stefanoel committed:
Update installer.xml

stefanoel committed:
utf-8 encoding

stefanoel committed:
Fixed null values check

The changes introduced bugs because some fields have 0 value option. So empty() is not suitable.
Used isset instead.

stefanoel committed:
code style fix

File must end with a newline character

Quy commented on administrator/components/com_installer/Model/ManageModel.php:
```diff
@@ -324,7 +324,7 @@ protected function getListQuery()
 		$core     = $this->getState('filter.core');
 
-		if ($status !== '')
+		if (isset($status) && $status !== '')
```
No need to check `isset` as these are assigned in lines 321-325.

stefanoel committed:
Merge branch '4.0-dev' into com_installer-manage

stefanoel committed:
Improved input check and code style fixes

ManageModel.php checks if filter values are null. In a ws call filter values are null instead of empty strings. Without this check the ws get a wrong response because the if statements are evaluated to true.

stefanoel committed:
Merge branch 'com_installer-manage' of https://github.com/stefanoel/joomla-cms-wstest into com_installer-manage

stefanoel committed:
Merge branch '4.0-dev' into com_installer-manage

Quy commented on administrator/components/com_installer/Model/ManageModel.php:
```diff
@@ -324,7 +324,7 @@ protected function getListQuery()
 		$folder   = $this->getState('filter.folder');
 		$core     = $this->getState('filter.core');
```
```suggestion
		$core     = $this->getState('filter.core', '');
```
Can it be done this way? If yes, then there is no need to check null later.

SharkyKZ replied:
This should already be an empty string. Except for client filter because it defaults to `null` here https://github.com/joomla/joomla-cms/blob/4c3bd6b93be32938f6dae43a48e109af4be17f8e/administrator/components/com_installer/Model/ManageModel.php#L81.

stefanoel replied:
I tested filter values.
Just after a call from **backend form** I get:
`$status: string '' (length=0)`
`$type: string '' (length=0)`
`$clientId: string '' (length=0)`
`$folder: string '' (length=0)`
`$core: string '' (length=0)`
**and the query is listing all the extensions:**
`string 'SELECT *,2*protected+(1-protected)*enabled AS status FROM #__extensions WHERE state = 0' (length=91)`

After a **ws call** I get
`$status: null`
`$type: null`
`$clientId: null`
`$folder: null`
`$core: null`
**and the query returns no results:**
`string 'SELECT *,2*protected+(1-protected)*enabled AS status FROM #__extensions WHERE state = 0 AND protected = 0 AND enabled = 0 AND client_id = 0 AND folder = ''' (length=159)`

@SharkyKZ it seems that populateState isn't reached calling a ws, therefore all filters are null.

`Folder` filter is similar to `type`, so I can check just `if ($folder)`, but for status, client_id and core I need a fix like @Quy suggested:
`$this->getState('filter.status','');`
`$this->getState('filter.client_id','');`
`$this->getState('filter.core','');`

stefanoel replied:
@SharkyKZ, I'm struggling on what you suggested, populateState isn't reached and I guess that this could be a bug... I noticed that pagination is not working at all. The links for pagination of every ws (not only this one) return the same result (the first 20 records)... even changing url params values I get the same results. Perhaps solving this problem and invoking populateState will avoid the checks on filters, like you suggested...

Quy commented on plugins/webservices/installer/installer.xml:
```diff
@@ -0,0 +1,19 @@
+<extension version="3.1" type="plugin" group="webservices" method="upgrade">
+    <name>plg_webservices_installer</name>
+    <author>Joomla! Project</author>
+    <creationDate>August 2017</creationDate>
```
Update?

stefanoel replied:
I copied it from other ws source, I thought it must be the same... I'll change version and creationDate

SharkyKZ commented:
`JsonApiView.php` needs to be renamed to `JsonapiView.php` and the class inside too.

stefanoel committed:
Filters check reviewed

Changed ManageModel.php filters check
Updated install.xml and JsonapiView.php

stefanoel committed:
Rename JsonApiView.php to JsonapiView.php

alikon commented:
thanks for your 1st contruibution 👍 
you need to add the sql files for the plugin install and update see #26850 as a reference

stefanoel committed:
Added sql installation and update files

stefanoel committed:
Merge branch '4.0-dev' into com_installer-manage

stefanoel committed:
Update 4.0.0-2019-10-29.sql

alikon commented:
I have tested this item :white_check_mark: successfully on 8e2d21a7f4afba3a326c064203f6be07cec625cc<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26765">issues.joomla.org/tracker/joomla-cms/26765</a>.</sub>

alikon commented:
tested with postman
![Screenshot from 2019-10-29 21-25-45](https://user-images.githubusercontent.com/181681/67806255-cd6aab80-fa92-11e9-9334-6eff7ca191d0.png)
and this is what it is exposed in the the app 
![Screenshot from 2019-10-29 21-27-09](https://user-images.githubusercontent.com/181681/67806360-ff7c0d80-fa92-11e9-9ae8-36e9ada8f528.png)

stefanoel committed:
Merge branch '4.0-dev' into com_installer-manage

nibra commented on administrator/components/com_installer/Model/ManageModel.php:
```diff
@@ -351,7 +351,7 @@ protected function getListQuery()
 		}
 
-		if ($folder !== '')
+		if ($folder)
```
Don't rely on automagic casts. Maybe you mean 
if (empty($folder))
? That covers null values as well.

SharkyKZ replied:
>Don't rely on automagic casts?

Why? The result should be the same as using `empty()`. Only difference `empty()` doesn't emit warning when variable is undefined.

nibra replied:
Because $folder is expected to be a string (or null), not a boolean. It's a matter of implicit documentation. Additionally, the hidden cast may change in future versions of PHP (although not very likely). The rule is: "Test what you want to test, not something that accidently behaves similar".

stefanoel replied:
Please, let me know if I have to use empty... is this topic closed? :)

SharkyKZ replied:
This isn't enforced by coding standards, both are used across the code. You can use what you want. `empty()` is just an unnecessary function call in this case.

nibra replied:
If you don't use empty() here, the code may fail one day (when/if PHP drops hidden casting).

SharkyKZ replied:
It will also fail when/if PHP removes `empty()` function (:

SharkyKZ commented on api/components/com_installer/Controller/ManageController.php:
```diff
@@ -0,0 +1,39 @@
+	 * The default view for the display method.
+	 *
+	 * @var    string
+	 * @since  3.0
```
```suggestion
	 * @since  __DEPLOY_VERSION__
```

SharkyKZ commented on api/components/com_installer/View/Manage/JsonapiView.php:
```diff
@@ -0,0 +1,55 @@
+		'version',
+		'folder',
+		'status',
+		'client_id'
```
```suggestion
		'client_id',
```

stefanoel committed:
Update api/components/com_installer/Controller/ManageController.php

Co-Authored-By: SharkyKZ <sharkykz@gmail.com>

stefanoel committed:
Update api/components/com_installer/View/Manage/JsonapiView.php

Co-Authored-By: SharkyKZ <sharkykz@gmail.com>

stefanoel committed:
Merge branch '4.0-dev' into com_installer-manage

zero-24 commented:
looks good to me. Than you for that awesome feature. @stefanoel would you mind fixing the conflicts so this can get a second test by me and than we can set this here RTC?

Please keep up the great work and sorry that it took that long to come back to you on this.

stefanoel committed:
Merge branch '4.0-dev' into com_installer-manage

Quy commented:
Please remove prefix `en-GB.` from language files.

Quy commented on administrator/components/com_admin/sql/updates/mysql/4.0.0-2019-10-29.sql:
```diff
@@ -0,0 +1,2 @@
+INSERT INTO `#__extensions` (`package_id`, `name`, `type`, `element`, `folder`, `client_id`, `enabled`, `access`, `protected`, `manifest_cache`, `params`, `checked_out`, `checked_out_time`, `ordering`, `state`) VALUES
```
Add to the installation sql.

stefanoel replied:
done, please check!

Quy replied:
Add to postgresql installation.

stefanoel replied:
sorry...

stefanoel committed:
Removed en-GB. prefix from language file name

stefanoel committed:
Merge branch '4.0-dev' into com_installer-manage

stefanoel committed:
Added plugin extension to the installation sql

Quy commented on plugins/webservices/installer/installer.xml:
```diff
@@ -0,0 +1,19 @@
+		<filename plugin="installer">installer.php</filename>
+    </files>
+	<languages>
+		<language tag="en-GB">language/en-GB/en-GB.plg_webservices_installer.ini</language>
+		<language tag="en-GB">language/en-GB/en-GB.plg_webservices_installer.sys.ini</language>
```
```suggestion
		<language tag="en-GB">language/en-GB/plg_webservices_installer.ini</language>
		<language tag="en-GB">language/en-GB/plg_webservices_installer.sys.ini</language>
```

Quy commented:
Add to the list of core extensions.

libraries/src/Extension/ExtensionHelper.php

stefanoel committed:
Added postgresql insert

stefanoel committed:
Added to the list of core extensions

stefanoel committed:
Fixed language file name

Quy commented on api/components/com_installer/Controller/ManageController.php:
```diff
@@ -0,0 +1,39 @@
+
+namespace Joomla\Component\Installer\Api\Controller;
+
+defined('_JEXEC') or die;
```
```suggestion
\defined('_JEXEC') or die;
```

Quy commented on api/components/com_installer/Controller/ManageController.php:
```diff
@@ -0,0 +1,39 @@
+	 * @since  __DEPLOY_VERSION__
+	 */
+	protected $default_view = 'manage';
+
```
```suggestion
```

Quy commented on api/components/com_installer/View/Manage/JsonapiView.php:
```diff
@@ -0,0 +1,55 @@
+
+namespace Joomla\Component\Installer\Api\View\Manage;
+
+defined('_JEXEC') or die;
```
```suggestion
\defined('_JEXEC') or die;
```

Quy commented:
I am getting a 500 Internal server error.

stefanoel committed:
Update api/components/com_installer/Controller/ManageController.php

Co-authored-by: Quy <quy@fluxbb.org>

stefanoel committed:
Update api/components/com_installer/View/Manage/JsonapiView.php

Co-authored-by: Quy <quy@fluxbb.org>

stefanoel committed:
Update api/components/com_installer/Controller/ManageController.php

Co-authored-by: Quy <quy@fluxbb.org>

alikon commented:
me too
![Screenshot from 2020-06-16 21-24-04](https://user-images.githubusercontent.com/181681/84819097-c5f31500-b017-11ea-8506-cf08fa2490b8.png)

stefanoel committed:
Merge branch '4.0-dev' into com_installer-manage

stefanoel committed:
Merge branch '4.0-dev' into com_installer-manage

stefanoel committed:
controller fixes

stefanoel commented:
> me too
> ![Screenshot from 2020-06-16 21-24-04](https://user-images.githubusercontent.com/181681/84819097-c5f31500-b017-11ea-8506-cf08fa2490b8.png)

I messed it up, it seems that files were on wrong location... now it works...

Quy commented:
I have tested this item :white_check_mark: successfully on 2f391305f00c1dfb550bfc3942de6ccb14e0e3b4<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26765">issues.joomla.org/tracker/joomla-cms/26765</a>.</sub>

alikon commented:
I have tested this item :white_check_mark: successfully on 2f391305f00c1dfb550bfc3942de6ccb14e0e3b4<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26765">issues.joomla.org/tracker/joomla-cms/26765</a>.</sub>

alikon commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26765">issues.joomla.org/tracker/joomla-cms/26765</a>.</sub>

wilsonge committed:
Simplify

wilsonge commented:
Thankyou very much for working so hard on this!

stefanoel commented:
> Thankyou very much for working so hard on this!



> Thankyou very much for working so hard on this!

Thank you all!!!
